src/integration.test.ts

```diff
@@ -58,65 +89,82 @@ describe("Integration Test", (): void => {
     });
   });
 
-  const mockedExec = async (load: boolean, command: string): Promise<void> => {
-    const stdout = "standard output from Bash command";
-    const stderr = "error output from Bash command";
-    child_process.exec.mockImplementationOnce(
-      (_command: any, _options: any, callback: any): any => {
-        callback(null, { stdout, stderr });
-      }
-    );
-
-    await (load ? docker.loadDockerImages() : docker.saveDockerImages());
-
-    expect(core.getInput).nthCalledWith(1, "key", { required: true });
-    expect(core.info).nthCalledWith<[string]>(1, command);
-    expect(child_process.exec).lastCalledWith(
+  const assertExecBashCommand = (n: number, command: string): void => {
+    const infoN = 2 * n - 1; // core.info is called twice for each time docker image list is called.
+    expect(core.info).nthCalledWith<[string]>(infoN, command);
+    expect(child_process.exec).nthCalledWith(
+      n,
       command,
       EXEC_OPTIONS,
       expect.anything()
     );
-    expect(core.info).lastCalledWith(stdout);
-    expect(core.error).lastCalledWith(stderr);
+    expect(core.error).lastCalledWith(STDERR);
     expect(core.setFailed).not.toHaveBeenCalled();
+    if (command === LIST_COMMAND) {
+      expect(core.info).nthCalledWith(2, dockerImages.join("\n"));
+    }
   };
 
-  test("cache misses, then hits", async (): Promise<void> => {
-    // Attempt first cache restore.
-    await docker.loadDockerImages();
+  const assertLoadDockerImages = (cacheHit: boolean): void => {
+    expect(cache.restoreCache).lastCalledWith([docker.DOCKER_IMAGES_PATH], KEY);
+    expect(core.saveState).nthCalledWith(1, docker.CACHE_HIT, cacheHit);
+    expect(core.setOutput).lastCalledWith(docker.CACHE_HIT, cacheHit);
+    if (cacheHit) {
+      expect(core.getInput).nthCalledWith(1, "key", { required: true });
+      assertExecBashCommand(1, loadCommand);
+    } else {
+      expect(core.setOutput).lastCalledWith(docker.CACHE_HIT, false);
+      expect(core.getInput).nthCalledWith(1, "key", { required: true });
+      assertExecBashCommand(1, LIST_COMMAND);
+      expect(core.saveState).lastCalledWith(
+        docker.DOCKER_IMAGES_LIST,
+        "test-docker-image:v1"
+      );
+    }
+  };
 
-    // Expect cache miss since cache has never been saved.
-    expect(core.getInput).nthCalledWith(1, "key", { required: true });
-    expect(core.setOutput).lastCalledWith(docker.CACHE_HIT, false);
+  const assertSaveCacheHit = (): void => {
+    expect(core.getInput).lastCalledWith("key", { required: true });
+    expect(core.info).lastCalledWith(
+      `Cache hit occurred on the primary key ${KEY}, not saving cache.`
+    );
     expect(child_process.exec).not.toHaveBeenCalled();
     expect(core.setFailed).not.toHaveBeenCalled();
+  };
+
+  const assertSaveCacheMiss = (): void => {
+    expect(core.getInput).nthCalledWith(1, "key", { required: true });
```

Comment:
This can be done once in `assertSaveDockerImages` rather than duplicated across branches.